### DISCUSSION TITLE
Code review, suggestions, ideas or feedback about my code

### DISCUSSION THREAD

**Discussion by msurdi** [statelyai/xstate#2487] [Show and tell]
Hi!, 

I'm just learning about XState and I'm really liking it so far. I've created a [side project](https://github.com/msurdi/videittous/) and would love any feedback regarding my use of XState on it. Would you do anything differently? How can I make it better? Am I doing things in an idiomatic way?

Sorry on beforehand if this doesn't belong here.

Thanks!

**Comment by ChrisShank** [comment]
With a quick glance everything looks really clean, but I'm not too familiar with the problem domain you are trying to model! Maybe to throw the question back at you, is there anything that you dont like or think could be improved? 🙃

**Reply by msurdi** [reply]
Hey! Thanks a lot for your time. The most confusing part to me was that, I have basically two machines: the first one controls the overall flow of the application, and when necessary it invokes another machine which takes care of some processing. This second machine can provide progress information, so what I'm doing is reporting this value via `sendParent` at almost every state in the child machine. [Example](https://github.com/msurdi/videittous/blob/7c6fed5b9d692e28f07a9c9275b81e9d25bcf9b2/src/machines/encodingMachine.js#L59).

I've read about the `sync` flag, but I understand that's just for spawned actors and not for invoked machines. Maybe I should move to a spawned actor instead of an invoke for this use case?

Again, thanks a lot for your help and time.

**Reply by ChrisShank** [reply]
Thanks for the reply! Take these comments more as food for thought, modeling is much more of an art than a science. I like the separation between the steps machine and the encoding machine as they both have a well defined behavior. It seems like the question is who should own the progress state. Instead of syncing the progress of the encoding machine with the steps machine you could use the encoding machine directly in the UI. This means that you should add `progress` to the context of encoding.

```ts
// Let react access the state of the invoked service
const [state, send] = useActor(state.children.encoding);
```


Nitpicks:
- [This](https://github.com/msurdi/videittous/blob/main/src/machines/stepsMachine.js#L69) `cond` is not necessary!
- Could be forgetting here but invoked callbacks generally should be async? Invoked callbacks let you return a cleanup function, so I'm sort of surprised that they allow you to return data from them! https://github.com/msurdi/videittous/blob/main/src/machines/encodingMachine.js#L102

**Reply by msurdi** [reply]
> const [state, send] = useActor(state.children.encoding);

I tried this (although I'm using Vue 3, but conceptually more or less the same thing can be accomplished...) the problem I found is that `state.children.encoding` does not exist when the hook is used. It will eventually exists as the top level machine invokes it. Maybe by wrapping this hook's usage in a component which is not rendered before the encoding machine is invoked might solve the issue, but not sure what's the better approach to be honest.

**Reply by ChrisShank** [reply]
Ya gotcha it can be a little tricky! Ill try to get a example up later today of how to deal with this (in Vue!)

**Reply by ChrisShank** [reply]
Ah sorry didn't realize you were using Vue! Typically the pattern I have use with `useActor` is as you just described:

```html
// EncodingPage.vue
<Progress v-if="state.children.encoding" :service="state.children.encoding">
```

```ts
// Progress.vue

export default defineComponent({
  setup(props) {
    const {state} = useActor(props.service);

    return {state}
  }
})
```

**Reply by github0013** [reply]
https://xstate.js.org/docs/guides/states.html#state-methods-and-properties

Just out of curiosity, but `state.children` seems to be a Map object when I `console.log(service)`.  And how does `state.children.encoding` (encoding should be the invoke id name) still works on a Map object?

```js
x = new Map([["encoding", "abc"]])
// => Map(1) {"encoding" => "abc"}
x.encoding
// => undefined
```

**Reply by ChrisShank** [reply]
This is a subtle discrepancy with xstate that is a little confusing (hopefully its resolved in the next breaking change):

```ts
state.children // regular object

service.children // Map
```

**Reply by github0013** [reply]
oh.. service and state..

And I think I found the explanation in the doc.
https://xstate.js.org/docs/guides/communication.html#referencing-services
> When JSON serialized, the state.children object is a mapping of service IDs (keys) to objects containing meta data about that service

**Reply by msurdi** [reply]
Thanks a lot for your explanations, this has been very useful to me💯

**Reply by msurdi** [reply]
So I just [followed all these recommendations](https://github.com/msurdi/videittous/commit/5e88f272fb625dfaf66477c6cf02adc8dd4a8f56) and the code looks a lot better to me now. Thanks!
